File: contracts/contract.go
// Package contracts is the single source of truth for the Vibe Forge API
// contract. It embeds contracts/contract.json (the same file the React frontend
// imports) via go:embed and exposes typed accessors so the backend can never
// drift from the frontend.
//
// The canonical contract is contracts/contract.json. Do not duplicate these
// values by hand elsewhere; add accessors here instead.
package contracts

import (
	_ "embed"
	"encoding/json"
	"fmt"
)

//go:embed contract.json
var raw []byte

// Contract is the parsed shared contract.
type Contract struct {
	Name         string             `json:"name"`
	Version      string             `json:"version"`
	Stages       Stages             `json:"stages"`
	States       map[string]Enum    `json:"states"`
	ArtifactTypes map[string]string `json:"artifactTypes"`
	Events       Events             `json:"events"`
	Paths        map[string]Path    `json:"paths"`
	Idempotency  Idempotency        `json:"idempotency"`
	Errors       Errors             `json:"errors"`
	Limits       Limits             `json:"limits"`
}

// Enum is a string state enumeration.
type Enum struct {
	Type   string   `json:"type"`
	Values []string `json:"values"`
}

// Stages defines the four serial agent-loop stages in order.
type Stages struct {
	Order        []string          `json:"order"`
	Descriptions map[string]string `json:"descriptions"`
}

// Events is the SSE event contract.
type Events struct {
	Protocol    Protocol                  `json:"protocol"`
	Definitions map[string]EventDef       `json:"definitions"`
}

// Protocol describes the SSE transport rules.
type Protocol struct {
	ContentType    string   `json:"contentType"`
	Seq            string   `json:"seq"`
	ReplayHeader   string   `json:"replayHeader"`
	TerminalEvents []string `json:"terminalEvents"`
}

// EventDef describes one SSE event.
type EventDef struct {
	Fields    map[string]string `json:"fields"`
	Semantic  string            `json:"semantic"`
}

// Path describes one REST endpoint.
type Path struct {
	Method  string            `json:"method"`
	Path    string            `json:"path"`
	Type    string            `json:"type,omitempty"`
	Success map[string]any    `json:"success,omitempty"`
	Errors  map[string]string `json:"errors,omitempty"`
}

// Idempotency describes the idempotency-key rules.
type Idempotency struct {
	Header    string   `json:"header"`
	TTLSeconds int      `json:"ttlSeconds"`
	AppliesTo []string `json:"appliesTo"`
}

// Errors is the stable error contract.
type Errors struct {
	Structure  map[string]string         `json:"structure"`
	Codes      map[string]ErrorCode       `json:"codes"`
	RunFailureCodes map[string]string     `json:"runFailureCodes"`
}

// ErrorCode maps a stable code to its HTTP status and retryability.
type ErrorCode struct {
	HTTP      int  `json:"http"`
	Retryable bool `json:"retryable"`
}

// Limits holds the shared numeric/string limits.
type Limits struct {
	PromptMinChars              int    `json:"promptMinChars"`
	PromptMaxChars              int    `json:"promptMaxChars"`
	FileContentMaxBytes         int    `json:"fileContentMaxBytes"`
	WritableFilePath            string `json:"writableFilePath"`
	RunWatchdogSeconds          int    `json:"runWatchdogSeconds"`
	SandpackReadyTimeoutSeconds int    `json:"sandpackReadyTimeoutSeconds"`
	MaxAutoFixRounds            int    `json:"maxAutoFixRounds"`
}

// cached is the parsed contract, populated on first use.
var cached *Contract

// Load parses and returns the embedded contract. It panics on parse failure
// because a corrupt contract is a compile-time-class bug that must surface
// immediately, never silently.
func Load() *Contract {
	if cached != nil {
		return cached
	}
	var c Contract
	if err := json.Unmarshal(raw, &c); err != nil {
		panic(fmt.Sprintf("contracts: failed to parse embedded contract.json: %v", err))
	}
	cached = &c
	return cached
}

// Raw returns the raw embedded contract bytes (for serving / debugging).
func Raw() []byte { return raw }

// Version returns the contract version string.
func Version() string { return Load().Version }

// HTTPStatusFor returns the HTTP status code for a stable error code.
// Returns 500 if the code is unknown.
func HTTPStatusFor(code string) int {
	if ec, ok := Load().Errors.Codes[code]; ok {
		return ec.HTTP
	}
	return 500
}

// IsRetryable reports whether a stable error code is retryable.
func IsRetryable(code string) bool {
	if ec, ok := Load().Errors.Codes[code]; ok {
		return ec.Retryable
	}
	return true
}

File: frontend/src/contract.ts
// Frontend view of the Vibe Forge shared contract. This module imports the
// canonical contracts/contract.json (the SAME file the Go backend embeds) and
// exposes typed constants. Never hand-copy values from the contract into other
// frontend files; import them from here so the frontend cannot drift from the
// backend.
import contract from "../../contracts/contract.json";

export const contractVersion: string = contract.version;

export const STAGES = contract.stages.order as readonly string[];
export type Stage = (typeof contract.stages.order)[number];

export const PROJECT_STATES = contract.states.project.values as readonly string[];
export const RUN_STATES = contract.states.run.values as readonly string[];
export const VERSION_STATES = contract.states.version.values as readonly string[];
export const STAGE_NODE_STATES = contract.states.stageNode.values as readonly string[];
export const WORKFLOW_PROJECT_STATES =
  contract.states.workflowProject.values as readonly string[];
export const WORKFLOW_STAGE_STATES =
  contract.states.workflowStage.values as readonly string[];

export type ProjectStatus = (typeof PROJECT_STATES)[number];
export type RunStatus = (typeof RUN_STATES)[number];
export type VersionStatus = (typeof VERSION_STATES)[number];
export type StageNodeStatus = (typeof STAGE_NODE_STATES)[number];
export type WorkflowProjectStatus = (typeof WORKFLOW_PROJECT_STATES)[number];
export type WorkflowStageStatus = (typeof WORKFLOW_STAGE_STATES)[number];

/** The 8 unified SSE event names (PRD-A/B/C §5.2). */
export const EVENT_NAMES = Object.keys(contract.events.definitions) as readonly string[];
export type EventName = (typeof EVENT_NAMES)[number];

export const IDEMPOTENCY_HEADER: string = contract.idempotency.header;
export const IDEMPOTENCY_TTL_SECONDS: number = contract.idempotency.ttlSeconds;

/** The only writable business file in the fixed scaffold (PRD-B §5.3). */
export const WRITABLE_FILE_PATH: string = contract.limits.writableFilePath;
export const PROMPT_MAX_CHARS: number = contract.limits.promptMaxChars;
export const PROMPT_MIN_CHARS: number = contract.limits.promptMinChars;
export const SANDPACK_READY_TIMEOUT_MS: number =
  contract.limits.sandpackReadyTimeoutSeconds * 1_000;

/** REST paths keyed by contract name (PRD-A/B/C §5.1). */
export const PATHS = contract.paths as Record<string, {
  method: string;
  path: string;
  type?: string;
}>;

/** Stable error codes -> HTTP status (contract §errors). */
export const ERROR_CODES = contract.errors.codes as Record<string, {
  http: number;
  retryable: boolean;
}>;

export type ApiError = {
  code: string;
  message: string;
  retryable: boolean;
  details?: Record<string, unknown>;
};

export default contract;
